lib/notifications/display.dart

```diff
@@ -202,6 +205,61 @@ class NotificationDisplayManager {
     );
   }
 
+  static void _onRemoveFcmMessage(RemoveFcmMessage data) async {
+    assert(debugLog('notif remove zulipMessageIds: ${data.zulipMessageIds}'));
+
+    final groupKey = _groupKey(data);
+
+    var haveRemaining = false;
+    final activeNotifications = await ZulipBinding.instance.androidNotificationHost
+      .getActiveNotifications(desiredExtras: [kExtraZulipMessageId]);
+    for (final statusBarNotification in activeNotifications) {
+      if (statusBarNotification == null) continue; // TODO(pigeon) eliminate this case
+      final notification = statusBarNotification.notification;
+
+      // Sadly we don't get toString on Pigeon data classes: flutter#59027
+      assert(debugLog('  existing notif'
+        ' id: ${statusBarNotification.id}, tag: ${statusBarNotification.tag},'
+        ' notification: (group: ${notification.group}, extras: ${notification.extras}))'));
+
+      // Don't act on notifications that are for other Zulip accounts/identities.
+      if (notification.group != groupKey) continue;
+
+      // Don't act on the summary notification for the group.
+      if (statusBarNotification.tag == groupKey) continue;
+
+      final lastMessageIdStr = notification.extras[kExtraZulipMessageId];
+      if (lastMessageIdStr == null) continue;
+      final lastMessageId = int.parse(lastMessageIdStr, radix: 10);
+      if (data.zulipMessageIds.contains(lastMessageId)) {
+        // The latest Zulip message in this conversation was read.
+        // That's our cue to cancel the notification for the conversation.
+        await ZulipBinding.instance.androidNotificationHost
+          .cancel(tag: statusBarNotification.tag, id: statusBarNotification.id);
+      } else {
+        // This notification is for another conversation that's still unread.
+        // We won't cancel the summary notification.
+        haveRemaining = true;
+      }
+    }
+
+    if (!haveRemaining) {
+      // The notification group is now empty; it had no notifications we didn't
+      // just cancel, except the summary notification.  Cancel that one too.
+      //
+      // Even though we enable the `autoCancel` flag for summary notification
+      // during creation, the summary notification doesn't get auto canceled if
+      // child notifications are canceled programatically as done above.
+      await ZulipBinding.instance.androidNotificationHost
+        .cancel(tag: groupKey, id: notificationIdAsHashOf(groupKey));
+    }
+  }
+
+  /// The identifier key for the message-id of embedded in the
+  /// notification's extras bundle.
+  @visibleForTesting
+  static const kExtraZulipMessageId = 'zulipMessageId';
```

Comment:
Thanks for adding this dartdoc. Does it have a missing word or something? I'm not parsing "[…] the message-id of embedded […]".

Also, "the notification" makes it sound like I should already know what kind of thing—and also which specific thing of that kind—is being referred to. Is it an instance of `StatusBarNotification`, or of `Notification`? Looking through the code, I think this string is used as a key in the `Notification.extras` map, so it seems helpful to include a reference to `[Notification.extras]`.

I'm not as familiar with the Android-notifications code as you and Greg are, and one question I had was whether this constant's value is part of the Zulip API, or if it's purely internal to our code. It seems like the latter, but I think if the dartdoc were fleshed out a bit more to answer that question, I think it would be helpful.